src/Functions/tuple.cpp

```diff
@@ -5,7 +5,18 @@ namespace DB
 
 REGISTER_FUNCTION(Tuple)
 {
-    factory.registerFunction<FunctionTuple>();
+    factory.registerFunction<FunctionTuple>(FunctionDocumentation{
+        .description = R"(
+Returns a tuple by grouping input arguments.
+
+For columns C1, C2, … with the types T1, T2, …, it returns a Tuple(C1 T1, C2 T2, …) type tuple containing these columns. There is no cost to execute the function.
+If there are duplicate column names, subsequent names will be suffixed with `_N` starting from zero. For example, `tuple(C1, C1)` will return a Tuple(C1 T1, C1_0 T1).
```

Comment:
> If there are duplicate column names, subsequent names will be suffixed with `_N` starting from zero. For example, `tuple(C1, C1)` will return a Tuple(C1 T1, C1_0 T1).

This doesn't sound very good. I mean such name would look quite random if that element of a tuple is not just a column in some table. I think maybe it's better instead to give the ability to specify names explicitly:

```
tuple(1 AS x, 2 AS y)
```

Reply:
It will be more ergonomic to allow named tuples from `tuple(x, y)` instead of `tuple(x AS x, y AS y)`.

About identical column names - what about not assigning any names when there is ambiguity?

Reply:
> It will be more ergonomic to allow named tuples from tuple(x, y) instead of tuple(x AS x, y AS y)
Yes, but it seems not very useful to generate name like `plus(number, 1)` for the second element of tuple `tuple(number, number + 1)`.

Reply:
And also for `SELECT tuple(3, 2, 1) AS t, t.1` there is ambiguity: should `t.1` mean `1` or `3`?

Reply:
> what about not assigning any names when there is ambiguity?

But we can't assign some names of a tuple and not assign the other names of the same tuple.
Maybe it makes sense to assign names like `element1`, `element2`, ... and so on for each element of a tuple which is not a literal name, or if that element is a duplicate of another element. There would be still ambiguity with `SELECT tuple(number, number + 1 AS element1) AS t, t.element1` (because both `number` and `number + 1` should be `element1`) but this case is tricky enough to just throw an exception that we can't deduct tuple names.

Reply:
What about using a named tuple only if all elements are named like proper identifiers?

Reply:
What if some of the elements are named like proper identifiers, and some are not?
Like

```
SELECT tuple(x, x + y, y) AS t, t.x, t.y
```

It seems convenient to allow that.

Reply:
In this case, we will not use names at all.

Reply:
(the user will have an option to define an alias for the missing column, or to do an explicit cast to a named tuple)

Reply:
Yes, this!

Reply:
What if a column is named as `1`? 
```
tuple(`1`, `2`)
```

Reply:
@amosbird, the tuple will be unnamed because `1` does not work as unquoted identifier.

Reply:
Numbers are dangerous as the names of tuple's elements because there could be confusion with accessing by index:
```
> SELECT CAST(('Hello', 'world') AS Tuple(`2` String, `1` String)) AS t, t.1, t.2, t.`1`, t.`2`

   ┌─t─────────────────┬─tupleElement(t, 1)─┬─tupleElement(t, 2)─┬─t.1───┬─t.2───┐
1. │ ('Hello','world') │ Hello              │ world              │ world │ Hello │
   └───────────────────┴────────────────────┴────────────────────┴───────┴───────┘
```

So let's at least not make such named tuples implicitly.

Reply:
I guess 
```
tuple(a AS `42`, b)
```
will also be unnamed then.

Reply:
Yes.

Reply:
> I guess
tuple(a AS `42`, b)
will also be unnamed then.

I think that should throw an exception instead of silently making an unnamed tuple `tuple(a, b)`. Because one name is specified and it doesn't feel right to just silently ignore a part of the command.